### PR TITLE
ICU-21569 Propagate LSTM memory allocation issue

### PR DESCRIPTION
**Pull request by FrankYFTang**
##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-21569
- [X] Required: The PR title must be prefixed with a JIRA Issue number. 
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [X] Issue accepted (done by Technical Committee after discussion)
- [X] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

**Comment by FrankYFTang** [comment]
Notice a big diff in BurmeseBreakEngine and KhmerBreakEngine in icu4c/source/common/dictbe.h is simply remove the ending trailing SPACEs.

**Comment by FrankYFTang** [comment]
Is there any suggestion how can I stress test the memory to force the uprv_malloc to fail? How can I add test to ensure the PR is "good enough" from the testing point of view beyond you sharp eyes of code review?

**Comment by aheninger** [comment]
Out of memory handling in RBBI is a known problem, and far more widespread than just the problems addressed by this  PR.

The issues are difficult to address because very few of the public RBBI API functions include an error code parameter, leaving no decent way to report errors when they occur, or to query for errors after-the-fact.

I think what needs to be done is, when a failure happens, to make sure the break iterator object is left in a failed-error state that wont cause future crashes, and to define & document what the API functions do with such an object. It may be necessary to add some new API for use by applications that are being careful in their handling of memory failures.

[ICU-21180](https://unicode-org.atlassian.net/browse/ICU-21180) is a closely related issue, for cleaning up error handling in the RBBI constructors. I think it should be broadened to cleaning up RBBI memory handling in general. I have some changes in progress for that, but it's really quite messy to clean up, and led to finding additional out-of-memory handling bugs with UVector.

**Comment by FrankYFTang** [comment]
The intend of this PR is to address the out of memory issue in my new LSTM code, not the whole RBBI code.

**Comment by FrankYFTang** [comment]
@jefgen  do you think this PR address the issue you have concern with?

**Review by jefgen — APPROVED** [review]
> @jefgen do you think this PR address the issue you have concern with?

Thanks for this PR! Yes, I think this address the concern I had on the other PR about `Array1D` not handling memory allocation failures. Thanks for adding a `status` parameter to the methods as well in order to propagate the error status.

Regarding holding off on these changes and/or using the other ICU ticket for OOM issues RBBI:
Thanks Andy! That is a good point that there are many other OOM issues in the RBBI code as well. I guess I'm somewhat on the fence about holding off on these changes specifically though, as this is new code that was just checked-in and it feels like we might want to try and check/handle OOM if we can. Since this is only changing files related to the recent LSTM changes and not holistically addressing OOM in other areas, perhaps using the LSTM ticket would be fine? (Sort of as a follow-up to the previous PRs). 

Since Frank already has a PR for this change, I don't know if I'd want to block it from being checked-in though. I'll defer to Frank if he wants to hold off on the change.

File: icu4c/source/common/lstmbe.cpp
Comment (FrankYFTang) [reply icu4c/source/common/lstmbe.cpp:384]: yes

File: icu4c/source/common/lstmbe.cpp
Comment (FrankYFTang) [reply icu4c/source/common/lstmbe.cpp:168]: but U_SUCCESS(status) I also need to call clear()

Comment (FrankYFTang) [reply icu4c/source/common/lstmbe.cpp:168]: but I also need to call clear() while U_SUCCESS(status)

File: icu4c/source/common/lstmbe.cpp
Comment (FrankYFTang) [reply icu4c/source/common/lstmbe.cpp:384]: yes, that is why

File: icu4c/source/common/lstmbe.cpp
Comment (jefgen) [reply icu4c/source/common/lstmbe.cpp:168]: Ah, good point. I think you are saying that you *don't* want to call `clear()` if `U_SUCCESS(status)` is false.
So the `clear()` should be inside the check for `U_SUCCESS(status)`. 
Thanks

File: icu4c/source/common/lstmbe.cpp
Comment (jefgen) [reply icu4c/source/common/lstmbe.cpp:384]: I meant that you don't need the check on line 381, since the next two lines will be no-ops if the status is failed. (But it's not a big issue to check the status there anyways).

Comment (FrankYFTang) [reply icu4c/source/common/lstmbe.cpp:384]: If line 379 fail and I didn't check on 381, then hunits_res.getAlias() could be nullptr, right? I try not to call ures_getInt with the first parameter as nullptr.

File: icu4c/source/common/lstmbe.cpp
Comment (FrankYFTang) [reply icu4c/source/common/lstmbe.cpp:168]: clear() would set memory, and if the status is fail, we should fail and return w/o perform any tasks.

Comment (jefgen) [reply icu4c/source/common/lstmbe.cpp:168]: Right, thanks!

File: icu4c/source/common/lstmbe.cpp
Comment (jefgen) [reply icu4c/source/common/lstmbe.cpp:384]: In `ures_getInt` it will first check if status is failed, or if the first argument is null. So if hunits_res.getAlias() is nullptr, then calling it should be fine. Though it might change the error status if it isn't already failed.
So maybe this is fine as-is.

**Pull request merged**
